salt/modules/cmdmod.py

```diff
@@ -443,12 +443,13 @@ def _run(
         # Ensure environment is correct for a newly logged-in user by running
         # the command under bash as a login shell
         try:
-            user_shell = __salt__["user.info"](runas)["shell"]
+            # Do not rely on populated __salt__ dict (ie avoid __salt__['user.info'])
+            user_shell = [x for x in pwd.getpwall() if x.pw_name == runas][0].pw_shell
             if re.search("bash$", user_shell):
                 cmd = "{shell} -l -c {cmd}".format(
                     shell=user_shell, cmd=_cmd_quote(cmd)
                 )
-        except KeyError:
+        except (AttributeError, IndexError):
```

Comment:
Can you elaborate why you changed the exceptions we are catching here?

Reply:
Sure. I assumed catching the KeyError was intended to handle `shell` not being defined in the return value of `user.info`. The new shell discovery relies on the returned list by getpwall() not being empty and the first item having a property of pw_shell, hence the changed exceptions. I might have missed `_cmd_quote` being called as well, not sure if that function is expected to throw a key error.